mon,mgr: extricate PGmap from monitor

liewegas opened the pull request:
Rebased https://github.com/ceph/ceph/pull/15022, then separated necessary bits of PGmap into PGMapDigest parent.

- [x] make check
- [x] robust flush_pg_stats machinery and/or adjust all of the tests that use it
- [x] qa rados
- [x] include PGMapDigest in 'mon report'
- [x] clean out pre-luminous pg stat queue on OSD (current code is racy)
- [x] limit size or remove health_detail from report
- [x] send pg scrub and osd scrub commands via mgr
- [x] trim old mgrstat states
- [x] clean up PGMapDigest encode/decode
- [x] clean up digestbl useless bufferlist wrapper

eventually
- [x] remove osd_stat from PGMapDigest (it gets big!)

gregsfortytwo commented on src/mon/OSDMonitor.cc:
```diff
@@ -922,10 +922,12 @@ OSDMonitor::update_pending_pgs(const OSDMap::Incremental& inc)
-  // should have prepared the latest pgmap if any
-  mon->pgservice->maybe_add_creating_pgs(creating_pgs.last_scan_epoch,
-					 &pending_creatings);
+  if (!osdmap.test_flag(CEPH_OSDMAP_REQUIRE_LUMINOUS)) {
```
If I read this correctly when updating, flagging this on LUMINOUS is exactly backwards. @tchaikov? I think we're trying to handle the overlap where the pgmap might have gotten an update which we didn't pick up in the separate creating_pgs set.

This also illustrates the main missing thing we need to deal with, which is the switch on upgrade from using the local pgmap to going via the ceph-mgr.
Also remaining: persistence of whatever state we need (from the mgr to the mon), and a sane way of dealing with stale data.

tchaikov replied:
no, i don't think it's backwards. we don't need to sync up with pgmap if all osds are luminous. but we need to do so while some osds are still pre-luminous. 

i try to keep `creating_pgs` updated with the bits from pgmap which is ensured to be updated **before** `CEPH_OSDMAP_REQUIRE_LUMINOUS` is set. so we can have a complete `creating_pgs` set when we are switched to `CEPH_OSDMAP_REQUIRE_LUMINOUS`. assuming following upgrade steps:

1. all mon are luminous, and `FEATURE_LUMINOUS` in monmap is set when the last luminous mon joins the quorum. but not all osds are luminous. so we can not set `CEPH_OSDMAP_REQUIRE_LUMINOUS`
1. the leader starts to update `creating_pgs`, but it is not able have the complete set of pgs being created just by consuming the (inc) osdmap after `FEATURE_LUMINOUS` is set. because

   1. there are some pools is still being created when the `FEATURE_LUMINOUS` is set. so the pgs belonging to that pools could be still being created also. and they only exist in pgmap.
   1. there are some osds are still at pre-luminous, so they keep `PGMonitor` (now `PGMapStatService`) updated with `MPGStats`, but they don't send `MOSDPGCreated` to mon.  that's what `trim_creating_pgs()` is for. please note we also guard it with `!osdmap.test_flag(CEPH_OSDMAP_REQUIRE_LUMINOUS)`  in `OSDMonitor::encode_pending()`. 
1. but once we have `CEPH_OSDMAP_REQUIRE_LUMINOUS` in osdmap, all osds are luminous. we don't need the pgmap for updating creating_pgs anymore, the luminous osds will ensure this.

am i missing anything?

gregsfortytwo replied:
Ah, that sounds right. The  dueling monitor and osdmap LUMINOUS flags just keep jumbling up in my head.

liewegas commented:
ok, added a bunch more stuff...

1. can't have the digest in MgrMonitor or else the epoch for MgrMap rolls over every 5 seconds. moved it to a new MgrStatMonitor.
2. Fixed the PGMap ref in PGMonitor

Now when i start with -o 'mon debug no require luminous = true' it works both before and after i 'ceph osd set require_luminous_osds'.

- still need to fix the upgrade thing @gregsfortytwo greg pointed out
- maybe the report frequency shoudl be every 1 or 2 seconds instead of every 5 seconds?
- mgr needs to do log INFO for pg stat updates so that you see the io stats etc in the log i think?

and obviously this whole series needs a lot of clean up...

tchaikov commented on src/osd/OSD.cc:
```diff
@@ -7458,6 +7467,13 @@ void OSD::_committed_osd_maps(epoch_t first, epoch_t last, MOSDMap *m)
+	newmap->test_flag(CEPH_OSDMAP_REQUIRE_LUMINOUS)) {
+      dout(10) << __func__ << " REQUIRE_LUMINOUS flag changed in " << newmap->get_epoch()
+	       << dendl;
+      clear_pg_stat_queue();
```
quote from https://github.com/liewegas/ceph/pull/35#discussion_r116745177

> this isn't quite sufficient since the pgs osdmaps lag and they will still queue themselves...

tchaikov replied:
@liewegas i don't think so. see https://github.com/ceph/ceph/pull/15073/commits/f71c18abbc4e6650ddbb0a72a25ea71697219e1f#diff-dfb9ddca0a3ee32b266623e8fa489626R2756 . please note, we are checking against the osdmap in OSD not `osdmap_ref` or `last_persisted_osdmap_ref`.

pg checks the osdmap flag before queuing themselves. imaging following sequence:

1. osdmap does not have `CEPH_OSDMAP_REQUIRE_LUMINOUS`, and pg enqueue itself 
1. osd consume an osdmap with `CEPH_OSDMAP_REQUIRE_LUMINOUS`, so the queue is cleared
1. pg does not enqueue itself anymore.

liewegas replied:
i'm assuming that the pg enqueuing itself is based on PG::get_osdmap(), which is often an older man than the OSD's OSDMap.

Perhaps what we need here is an OSDService::get_min_pg_epoch() before/after comparison and when *that* crosses the flag threshold then we clear the queue?

tchaikov replied:
@liewegas , pg enqueues itself based on `PG::get_osdmap()`, but see https://github.com/ceph/ceph/commit/f71c18abbc4e6650ddbb0a72a25ea71697219e1f#diff-dfb9ddca0a3ee32b266623e8fa489626R2756, i disable the enqueue based on the OSD's latest osdmap.

tchaikov commented:
@liewegas could you elaborate a little bit 

> include PGMapDigest in 'mon report'

? what is "mon report"? maybe i can work on it tomorrow.

> regarding to `flush_pg_stats`

maybe we can have a helper in `ceph-helper.sh`, and it calls `osd flush_pgs_stats` and then `mgr send-mon`?

liewegas commented:
oh i see!  ok

liewegas commented:
@tchaikov look for "report" in Monitor.cc.. it's a big json dump of everything interesting about the cluster state.

tchaikov commented:
thanks. i will work on it tomorrow if nobody takes it by then.

tchaikov commented:
https://github.com/liewegas/ceph/pull/36 is being tested at http://pulpito.ceph.com/kchai-2017-05-17_14:46:45-rados-wip-mgr-stats-kefu---basic-smithi/

tchaikov commented:
being tested at http://pulpito.ceph.com/kchai-2017-05-20_17:29:58-rados-wip-bigbang---basic-mira/

tchaikov commented:
being tested at http://pulpito.ceph.com/kchai-2017-05-21_04:27:34-rados-wip-bigbang---basic-mira/

tchaikov commented:
being tested at http://pulpito.ceph.com/kchai-2017-05-27_07:55:19-rados-wip-mgr-stats-kefu---basic-mira/, seems all qa/workunits/rados/test.sh failed. but the teuthology.log is too large (1.6GB) to inspect. will find a way to shrink it a little bit.

tchaikov commented:
will rebase this branch against master tmr.

tchaikov commented:
the rados qa run: http://pulpito.ceph.com/kchai-2017-05-29_07:20:22-rados-wip-mgr-stats-kefu---basic-mira/

seems http://pulpito.ceph.com/kchai-2017-05-29_07:20:22-rados-wip-mgr-stats-kefu---basic-mira/1241403/ can be reproduced. i will take a look at it tomorrow.

tchaikov commented:
@sage how shall we "remove osd_stat from PGMapDigest", are we going to move features based on this stats to mgr? for example,  `OSDMonitor` uses it for preparing the output of `osd df`. if that's the case, i can do it tomorrow.

liewegas commented:
I don't think the osd_stats are a blocking issue.  And @gregsfortytwo already started on it I believe.  If this is passing tests we should merge

tchaikov commented:
i see. http://pulpito.ceph.com/kchai-2017-05-29_18:22:20-rados-wip-mgr-stats-kefu---basic-mira/ still has 17 test failed/timedout. most of them include `rados.yaml`. i will fix them before the merge.

gregsfortytwo commented:
Yep, working on the osd_stat map.

Kept getting blocked on other things so I haven't been able to track this the way I'd like, should be able to do so tomorrow. Does upgrading work properly, and been tested?

tchaikov commented:
@gregsfortytwo i thought the rados suite run should cover some upgrade tests, but it turns out that my last rados run does not include any of them: http://pulpito.ceph.com/kchai-2017-05-27_07:55:19-rados-wip-mgr-stats-kefu---basic-mira/?suite=rados%3Aupgrade.

i am scheduling a upgrade run at http://pulpito.ceph.com/kchai-2017-05-30_05:52:48-upgrade-wip-mgr-stats-kefu---basic-mira/ . let's see how it goes!

gregsfortytwo commented on src/mon/OSDMonitor.cc:
```diff
@@ -1073,14 +1130,17 @@ void OSDMonitor::prime_pg_temp(
 {
-  if (mon->monmap->get_required_features().contains_all(
+  if (mon->monmap->get_required_features().empty()) {
+    dout(10) << __func__ << ": quorum not formed" << dendl;
```
@tchaikov, don't we have a way to directly query the quorum status, instead of this indirect feature query?

liewegas replied:
i think this should just be dropped, actually.. the monmap doesn't know anything about quorum, and i can't tell what the intent was here

gregsfortytwo replied:
It's trying not to read the data if we don't expect it to be valid. I'm not sure if that's necessary or useful — I think I said elsewhere that it seemed weird for us to be invoking any of these functions while not active...

gregsfortytwo commented on src/mon/OSDMonitor.cc:
```diff
@@ -1434,24 +1494,29 @@ void OSDMonitor::share_map_with_random_osd()
 {
-  epoch_t floor;
+  if (mon->monmap->get_required_features().empty()) {
+    dout(10) << __func__ << ": quorum not formed" << dendl;
```
@tchaikov, same here about direct quorum querying. We can query ```!Monitor::get_quorum().empty()``` for instance.

gregsfortytwo commented on src/mon/PGMonitor.cc:
```diff
@@ -535,6 +578,10 @@ bool PGMonitor::preprocess_query(MonOpRequestRef op)
 bool PGMonitor::prepare_update(MonOpRequestRef op)
 {
+  if (mon->osdmon()->osdmap.require_osd_release >= CEPH_RELEASE_LUMINOUS) {
+    return false;
```
@liewegas, there's some haphazard quick returns on invoking pgmon post-luminous, but shouldn't we be doing something like removing it from the dispatch loops entirely?
I'd actually think we want to just not load it into memory at all once we're past the upgrade point, and maybe even unload it, but I haven't looked into how tricky that is yet....

liewegas replied:
There is no 'loop' per se, just a couple call sites in Monitor.cc.  We can move these calls to the caller but I don't think it makes things any better or worse.  I don't know that there is much more we can do to remove/disable this code until we don't have to support the upgrade case any more (or that we'd improve runtime performance by doing so).

gregsfortytwo replied:
I meant the call sites in the dispatch event loop. :)

And speaking of those call sites, see below about disabling it.

gregsfortytwo commented on src/mon/OSDMonitor.cc:
```diff
@@ -395,14 +405,22 @@ void OSDMonitor::update_from_paxos(bool *need_bootstrap)
     }
     if (mon->monmap->get_required_features().contains_all(
           ceph::features::mon::FEATURE_LUMINOUS)) {
-      creating_pgs = update_pending_pgs(inc);
```
@tchaikov @liewegas did this ever get discussed? From commit:

> We are not persisiting the updated creating_pgs here; this is bad!  I'm not sure why it was there to begin with?

It originated in 43ded67243c88a1abb7f07aa4dbdc326eec7633a as its own hunk, so it was clearly deliberate. I think it must have been so we could identify some of the state-changed PGs and I'm a bit worried about dropping the hunk. (For instance, if we've deleted a pool with creating pgs, we don't want to tell OSDs to create them! And we don't need to persist that data since any subsequent load will produce the same outcome.)

liewegas replied:
It probably made sense at the time, but it is incorrect to call update_pending_pgs() and not persist the result.  @tchaikov can you confirm this isn't needed?

tchaikov replied:
@gregsfortytwo i think we can drop this line.

if we just deleted a pool with creating pgs, the quorum should be aware of this and `creating_pgs` should be updated. so it should not contain the creatings pgs in that pool. if a new mon just joins the quorum, it will be sync'ed with another monitor. `OSDMonitor::get_store_prefixes()` returns both "osdmap" and "osd_pg_creating", so "pg_creating" will be sync'ed. this ensures us that the new joiner will also have an updated copy of `creating_pgs`.

gregsfortytwo commented on src/mon/PGMonitor.cc:
```diff
@@ -419,15 +444,36 @@ void PGMonitor::apply_pgmap_delta(bufferlist& bl)
+    }
+    put_last_committed(t, pending_inc.version);
+    put_value(t, "deleted", 1);
+    return;
```
@liewegas, I don't remember how much we care about the pgmap state, but should we perhaps copy it elsewhere in case of disaster?

liewegas replied:
We could leave it where it is, but then I don't know when we *would* delete it. My thinking is that since we don't bother storing it now it won't hurt to throw out the legacy info.  We won't start trimming again until we reestablish a new last_epoch_clean lower bound, so what would we use it for?  Maybe we could pg dump to the log or something?

gregsfortytwo replied:
I was thinking more disaster recovery. But yeah, I'm fine with tossing it. Just wanted to check.

gregsfortytwo commented on src/mon/PGMonitor.h:
```diff
@@ -32,16 +32,15 @@ using namespace std;
+  std::unique_ptr<PGStatService> pgservice;
+
+  bool do_delete = false;   ///< propose deleting pgmap data
+  bool did_delete = false;  ///< we already deleted pgmap data
```
ah, we've got some haphazard did_delete checks in the PGMonitor. How about instead we expose that to the Monitor via an is_enabled() interface, and crash out if anybody tries to access a PaxosService which is not enabled?

liewegas replied:
It's a one-time event so I'm not sure getting Monitor involved will help.  I'm also trying to keep the hackery confined to PGMonitor as much as possible.

I like the idea of making it crash out, though.. maybe pgmon() could return null once the switch is flipped?

gregsfortytwo replied:
Yeah, I was just trying to be general about it. Simple hacks are simple, though!

gregsfortytwo commented:
@liewegas, did these commits sneak in from your messenger lossy connections branch?

"mgr/DaemonServer: keep registry of osd sessions
Occasionally we send them messages."
"mgr/DaemonServer: use registered osd session to send scrub messages
We can't rely on the messenger to track lossy connections for us."

gregsfortytwo commented on src/mon/MgrStatMonitor.cc:
```diff
@@ -0,0 +1,200 @@
+  dout(10) << " " << version << dendl;
+  bufferlist digestbl, bl;
+  ::encode(pending_digest, digestbl, mon->get_quorum_con_features());
+  ::encode(digestbl, bl);
```
@liewegas, the commit adding this digestbl says it's "just for bigbang's benefit", although I'm not sure how. What's up, and should it actually go int?

liewegas replied:
Yeah he encoding into a bufferlist that then gets encoded again is there because the first instance of bigbang did that and had to maintain compat. Saem goes for PGMapDigest, which has somewhat weird encoding.  I'll separate those patches out

gregsfortytwo commented on src/mon/Monitor.cc:
```diff
@@ -4707,6 +4715,72 @@ int Monitor::print_nodes(Formatter *f, ostream& err)
+      continue;
+    reply->pool_stats[pool_name] = *pool_stat;
+  }
+  send_reply(op, reply);
```
@tchaikov, we've managed to keep the RADOS cluster stuff out of Monitor.cc so far; do you think we can put these stat handlers in OSDMonitor?

tchaikov replied:
@gregsfortytwo yes, i agreed.

tchaikov replied:
after a second thought, i think they should go to `MgrStatMonitor`. `MgrStatMonitor` is "closer" to where these stats requests are served in long term than `OSDMonitor`.

gregsfortytwo replied:
Don't forget we need to route it correctly before the upgrade is complete.

tchaikov replied:
ahh, my patch didn't get into this one. it does take this into consideration.

gregsfortytwo commented:
Okay, not a full review but I skimmed through it enough to feel good about what's gone on. :) Notably I did *not* check the pg temp rework. Let me know if somebody hasn't looked at that yet.

Now, let me rebase some patches and get that osd_stat out of there...

tchaikov commented:
@gregsfortytwo the pg temp changes is being reviewed at #15291. it think it's correct per se. but could be improved a little bit. maybe it can use more reviews.

liewegas commented:
Good catch, i'll strip out those stray commits, thanks!

liewegas commented:
partly this is a confusing log message.  empty get_require_features() means a full kraken or luminous quorum has never been formed.

gregsfortytwo commented:
Oh, tried to build some patches on top of this and 92d49f941e9bf0f1b9df7fb4f9e6c0dda2dfca7d is broken (it's referring to an osd_con member which I think you undid from the messenger lossy cons thing).

gregsfortytwo commented:
I also looked into it and at this point the only programmatic access I'm noticing is to pool stats requests. Stale returns or going back in time there can have some nice cascade consequences on eg Cinder, but I don't think it requires any serious code on our end. At this point, there are only two non-constant size members of the report: the osd_epoch vector and the health outputs.

So some remaining things I was considering:
* make the ceph CLI/rest modules request info from the mon and the mgr, so that we don't need to route anything to the monitor (except for backwards compatibility and answering the pool stats queries)
* (alternative: forward all requests to the manager so that it serves up the info and the monitor just provides a fallback when the mgr is dead?)
* remove the MgrStatsMon; its only purpose is persisting useless data and providing a separate "channel" for the pg health messages
* log a pgmapdigest to the monitor config-key store once a minute or something, so we don't have to go totally stale on manager failover but also don't need special machinery for it.
* (other alternative: if we really prefer the existing state of affairs, we can at least only persist the pgmapdigest once a minute or something [by handling them in preprocess instead of prepare]; that way we don't need to limit the number of pg health messages we provide)

Oh, and one thing we definitely should do: reset the PGMapDigest encode/decode functions. There's no reason to start them at version 6 and have already-dead members on our first release. ;)

tchaikov commented:
rebased at https://github.com/ceph/ceph/pull/15128

liewegas commented:
- added todo to clean up encodings before final merge (keeping this in place for now just for bigbang's benefit so we don't have to wipe the cluster; jsut a few days left)
- I'm leaning toward not persisting any pgmapdigest for the mgr because the user-visible effects will be more confusing.  Currently you either get (1) a recent value that never goes back in time or (2) stale.  If we persist at semi-regular intervals, then the observed values may go back in time in a confusing way. Also, if we persist at a different interval than the current mgrstats digest on the mon, then the pg flush sequence numbers won't be useful for determining if the stats you are seeing are up to date or not.
- I think it's important that ceph health return without a call-out to the mgr, which means the mon has to stash a recent report of health values etc.  The only alternative that makes sense is a timeout + fallback on every request but what we have now seems simpler!

gregsfortytwo commented:
That sounds fine to me.

Note that the build is still broken... (https://github.com/ceph/ceph/pull/15073#issuecomment-305357969)

tchaikov commented:
http://pulpito.ceph.com/kchai-2017-06-02_10:43:35-rados-wip-mgr-stats-kefu---basic-mira/1256577/

is the last rados qa failure we need to fix. i am able to reproduce it this afternoon. but didn't have the time to root cause it yet.

 master is good: http://pulpito.ceph.com/kchai-2017-06-02_14:13:42-rados-wip-badone-testing-2---basic-smithi/ (i tested @badone 's branch, assuming it is almost the same as master)

tchaikov commented:
will take a look at the upgrade suite: http://pulpito.ceph.com/kchai-2017-06-03_06:44:51-upgrade-wip-mgr-stats-kefu---basic-mira/. quite a few of them failed.

tchaikov commented:
@liewegas @gregsfortytwo  http://pulpito.ceph.com/kchai-2017-06-04_06:03:05-upgrade-master---basic-mira/: i just ran the master branch through our upgrade suite. it appears that wip-mgr-stats and master perform equally bad. i think this branch is ready to merge once we have https://github.com/liewegas/ceph/pull/50 merged and squashed. what do you think?

liewegas commented:
\o/
